### PR TITLE
Implement WebSocketConnection class

### PR DESCRIPTION
**Pull request by yummyblabla**
Included jest-websocket-mock and mock-socket for jest testing WebSocket connections.

Created WebSocketConnection class as a proxy for interacting with websocket object.
Methods
- sendMessage
- addListener
- removeListener

Created SendMessageValidator class to validate client-sent messages for code consistency

### CODE REVIEW

File: src/modules/ws/websocket.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+const WEBSOCKET_URL = process.env.REACT_APP_WEBSOCKETURL || '';
+
+/**
+ * Interface that the client can use for Websocket Connection.
+ */
+class WebSocketConnection {
+  public constructor(url: string = WEBSOCKET_URL) {
+    this.ws = new WebSocket(url);
+    this.ws.onopen = (event) => {
```
Comment (bcitcstkevm) [inline src/modules/ws/websocket.ts:16]: suggest onOpen?

File: src/modules/ws/websocket.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      console.log(event);
+      this.ws.send('');
+    };
+
+    this.ws.onclose = (event) => {
```
Comment (bcitcstkevm) [inline src/modules/ws/websocket.ts:21]: same as above camel case?

File: src/modules/ws/websocket.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      console.log(event);
+    };
+
+    this.ws.onerror = (error) => {
+      console.log(error);
```
Comment (bcitcstkevm) [inline src/modules/ws/websocket.ts:26]: suggest we build a common logger. has info related to given build

Comment (bcitcstkevm) [reply src/modules/ws/websocket.ts:26]: maybe trello todo
or leave todo line

File: src/modules/ws/websocket.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    this.ws.onmessage = (event) => {
+      console.log(event);
+    };
+  }
+}
+
+/**
+ * Sends a message to the backend if the message is validated.
+ */
+WebSocketConnection.prototype.sendMessage = function sendMessage(key: string, data: Record<string, unknown>): boolean {
```
Comment (bcitcstkevm) [inline src/modules/ws/websocket.ts:38]: I've never seen a function bound to the prototype, could you elaborate on using prototype to hold function?

File: docs/Mahjong UML.drawio [hunk 5]
```diff
@@ -1 +1 @@
...
...azyqJe33LxHRhni9poD5P/UBTPk0R+AGi1SKVBgIM/1x8lrHgSCz4eAm/ruhGqTj2x4zIYsODHA9q4EDrS8gMmfStEd6215AZMunbrn+3vogszdntMOAKB5hIuVdlBkd5p9yrLYNjiHZRzIb2LQ3kyH7c87du9HHUncFL4smRvSJ4287kGKKFXHJkb4YArWdyjGIur4+DVShH9gpH9u3u2aeTyDFEs/ZSiRxDJnKOhQKzNElOZai3RPW6TO1tQ4NUe3VQoP08Du82UKn2KoK39TyOKHkv1V4zBGg9j9M9VLeTtwS2UOLLb0ljcJThBU8ZpnuuVs8RUZQY3j10ivf1OPCvvKWnZMjoVlLe4/GhtpBhFlO/CR36vZv+3R2HEVN7McVYkqI+UvB0YrOkECWGB1e9+97gYXTP4cXMtYk9i+J9ciU1aqIGT2M2S42ixMjVCKTCaFJhdM+PKwyNt4ebptQ1DzFFCmPUm/AkBvUbGTPKxgyzCo3BZ0RtWy2LNMbTiEuI2CCSELURgqMvGiaESF/c3t0/XHEYEdB7cCQl6qMER1c0TAlRdnpy95VLiRBGdkmJGinBe+xLw5wQla/vR8N/hve98d0thxgEOUNiz+nuA5IctZGDs0lNw+QQ3aXydPXnw0hMj3hbZUmQ2gnSbV11ikrg4/vR6FZMkDlBKJAEqZ0gVtsq1CoOL7kCuExlNLqMxjyeymh4GY0lGmQmo96EGz1C8B0ZOMoHjq0TfqCFNKL81uX13RM3NU7NLvPidZKi9aU1Fk9t5FCWS2veC2/bS2ss0f75cmlNMwRodGnNZmo9rVbutzPtsm/+VDTycpY+QKONbYRyO/wc31eI9/s5z+Li9uvyYah+E6EDP3L/+XyT+Pl8g+3TOKXezvR2wYE45BDq7TM9t27dSnX0jpcZvKcPVnJHGo8AokLRmj2jZkuH3zzCHsFd7FxvUdVc6GsLsLyS0O+pn6oHl6OO+X5dE7bCncJQ4LCnrn56RpvEx8EydO8w8AxNccDISes9mmHipHmY3T6r4DnA6yBNx3zOnmWE7N99B+LqqcS5o63ZOFFM2k0yFj0CXsAZQFCKhKpEQm6XIlVROBTgrUGr5PkEPAqIknL5bC2AADgW87X0nZQmSMaeE4s9gtnHAY1S0eJHOCSY5ua258bEXro32EG0x/8B</diagram></mxfile>
```
Comment (bcitcstkevm) [inline docs/Mahjong UML.drawio:1]: new line :P

File: src/modules/ws/websocket.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    };
+
+    this.ws.onerror = (error) => {
+      console.log(error);
+    };
+
+    this.ws.onmessage = (event) => {
+      console.log(event);
+    };
+  }
```
Comment (jzc1828) [inline src/modules/ws/websocket.ts:32]: maybe put sendMessage function as a class method inside the `WebSocketConnection` class?
```js
sendMessage(key: string, data: Record<string, unknown>): boolean {
  ...
}
```

File: src/modules/ws/websocket.ts (hunk 4 above)
Comment (jzc1828) [reply src/modules/ws/websocket.ts:38]: prototype basically holds fields and methods of a js class, but the fields and methods can be modified even after a class has been instantiated. You can try it in the browser lol. Note: ts classes are eventually compiled to prototypes, but it is easier to read and safter (can add access modifiers, public private etc.)

Comment (bcitcstkevm) [reply src/modules/ws/websocket.ts:38]: yeah but why would you want to modify a method after a class has been instantiated? and for fields isn't that what gettings and setters are for?

Comment (bcitcstkevm) [reply src/modules/ws/websocket.ts:38]: Oh, it allows for the interface. I see what you did.

Comment (bcitcstkevm) [reply src/modules/ws/websocket.ts:38]: That's interesting. What is the value of such an interface?

File: src/modules/ws/websocket.ts (hunk 1 above)
Comment (yummyblabla) [reply src/modules/ws/websocket.ts:16]: Unfortunately, websocket methods are not camelcase
https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

File: src/modules/ws/websocket.ts (hunk 2 above)
Comment (yummyblabla) [reply src/modules/ws/websocket.ts:21]: Same as above

File: src/modules/ws/websocket.ts (hunk 3 above)
Comment (yummyblabla) [reply src/modules/ws/websocket.ts:26]: Okies

File: src/modules/ws/websocket.ts (hunk 1 above)
Comment (bcitcstkevm) [reply src/modules/ws/websocket.ts:16]: ahh IC

File: src/modules/ws/websocket.ts (hunk 4 above)
Comment (yummyblabla) [reply src/modules/ws/websocket.ts:38]: Interface allows for control of the object that we have created. User would not have direct access to the websocket object and will need to use our interface to use its functionality. https://refactoring.guru/design-patterns/proxy

File: docs/Mahjong UML.drawio (hunk 5 above)
Comment (yummyblabla) [reply docs/Mahjong UML.drawio:1]: drawio file. This is not code

File: src/modules/ws/websocket.ts (hunk 6 above)
Comment (yummyblabla) [reply src/modules/ws/websocket.ts:48]: Done

**Pull request merged**